recipe-client-addon/lib/Sampling.jsm

```diff
@@ -130,4 +130,35 @@ this.Sampling = {
 
     return Sampling.isHashInBucket(inputHash, wrappedStart, end, total);
   }),
+
+  /**
+   * Sample over a list of ratios such that, over the input space, each ratio
+   * has a number of matches in correct proportion to the other ratios.
```

Comment:
> Is that right?

Yep. I ran some tests in the browser console to confirm the behavior:

<img width="853" alt="screen shot 2017-03-20 at 11 56 51 pm" src="https://cloud.githubusercontent.com/assets/193106/24208817/d94760e6-0ee1-11e7-8738-559441593d18.png">

> Also, I wonder if this would be simpler as its own PR - it doesn't seem strictly related to the preferences API?

Yeah, in retrospect that's a good cut point. At the time what I did was write the action code, then write the system add-on APIs to support it, and choosing a branch seemed critical.